File: CloudBackup.Common/JobOptions.cs

namespace CloudBackup.Common
{
    public class JobOptions
    {
        public RetentionPolicy RetentionPolicy { get; set; }

        public ReplicationOptions ReplicationOptions { get; set; }

        public InstanceBackupMode InstanceBackupMode { get; set; }

        public BackupAfterErrorStatus StatusAfterScriptError { get; set; } = BackupAfterErrorStatus.Warning;

        public BackupAfterErrorStatus StatusAfterReplicationError { get; set; } = BackupAfterErrorStatus.Failed;

        public JobOptions()
        {
            RetentionPolicy = new RetentionPolicy();
            ReplicationOptions = new ReplicationOptions();
            TagJoinType = TagJoinType.Or;
        }

        public TagJoinType TagJoinType { get; set; }
    }

    public enum BackupAfterErrorStatus
    {
        Failed = 0,
        Warning = 1
    }

    public enum TagJoinType : byte
    {
        And = 0,
        Or = 1
    }

    public enum InstanceBackupMode
    {
        Default = 0, // "Snapshots + Initial AMI" for Windows, "Snapshots Only" for Linux
        [Description("AMI Only")]
        AmiOnly = 1
    }

    public class ReplicationOptions
    {
        public bool EnableReplication { get; set; }

        public int BackupInverval { get; set; }

        public List<string> ReplicationRegions { get; set; }

        public ReplicationOptions()
        {
            ReplicationRegions = new List<string>();
        }
    }

    public class RetentionPolicy
    {
        public RetentionPolicy()
        {
            RestorePointsToKeep = 10;
            TimeIntervalValue = 14;
            RetentionTimeIntervalType = RetentionTimeIntervalTypes.Days;
            RetentionAction = RetentionAction.RemoveBackup;
        }

        public TimeSpan GetTimeSpan()
        {
            if (RetentionTimeIntervalType == RetentionTimeIntervalTypes.Days)
            {
                return TimeSpan.FromDays(TimeIntervalValue);
            }
            else if (RetentionTimeIntervalType == RetentionTimeIntervalTypes.Hours)
            {
                return TimeSpan.FromHours(TimeIntervalValue);
            }
            else if (RetentionTimeIntervalType == RetentionTimeIntervalTypes.Months)
            {
                return TimeSpan.FromDays(TimeIntervalValue * 31);
            }

            return new TimeSpan();
        }

        public int RestorePointsToKeep { get; set; }

        public int TimeIntervalValue { get; set; }

        public RetentionTimeIntervalTypes RetentionTimeIntervalType { get; set; }

        public RetentionAction RetentionAction { get; set; }
    }

    public enum RetentionAction
    {
        RemoveBackup = 0,
        ArchiveBackup = 1
    }

    public enum RetentionTimeIntervalTypes
    {
        Days = 0,
        Hours = 1,
        Months = 2
    }
}

File: CloudBackup.Management/ClientApp/src/app/classes/JobOptions.ts
import { RetentionPolicy } from "./RetentionPolicy";
import { ReplicationOptions } from "./ReplicationOptions";

export enum BackupAfterErrorStatus {
    Failed = 0,
    Warning = 1
}

export enum InstanceBackupMode {
    Default = 0,
    AmiOnly = 1
}

export enum TagJoinType {
    And = 0,
    Or = 1
}

export class JobOptions {
    AppConsistent = false;
    NoReboot = false;
    RetentionPolicy: RetentionPolicy;
    ReplicationOptions: ReplicationOptions;

    StatusAfterScriptError!: BackupAfterErrorStatus;
    StatusAfterReplicationError!: BackupAfterErrorStatus;

    InstanceBackupMode!: InstanceBackupMode;
    TagJoinType!: TagJoinType;

    constructor(retentionPolicy: RetentionPolicy, replicationOptions: ReplicationOptions) {
        this.RetentionPolicy = retentionPolicy;
        this.ReplicationOptions = replicationOptions;
    }

    static Copy(obj: any): JobOptions {
        var policy = RetentionPolicy.Copy(obj.RetentionPolicy || obj.retentionPolicy);
        var replication = ReplicationOptions.Copy(obj.ReplicationOptions || obj.replicationOptions);
        let copy = new JobOptions(policy, replication);
        copy.AppConsistent = obj.appConsistent || obj.AppConsistent || false; // for backwards compatibility
        copy.NoReboot = obj.noReboot || obj.NoReboot || false; // for backwards compatibility

        copy.StatusAfterScriptError = obj.statusAfterScriptError || obj.StatusAfterScriptError;
        copy.StatusAfterReplicationError = obj.StatusAfterReplicationError || obj.StatusAfterReplicationError;

        copy.InstanceBackupMode = obj.InstanceBackupMode || obj.InstanceBackupMode;
        copy.TagJoinType = obj.TagJoinType;

        return copy;
    }
}
